0029-loopback-transport/0029-loopback-transport.md

```diff
@@ -0,0 +1,33 @@
+---
+# Loopback Transport Protocol (LT)
+
+The Loopback Transport protocol is a technique for transporting value over the Interledger. It relies on the establishment of a temporary, direct, ILPv4 link between sender and receiver (the loopback link). The flow is as follows.
+
+## Flow
+
+The following flow describes how the sender's loopback transfer module ("LT module") interacts with the application layer protocol, the receiver, and the ILPv4 network.
+
+1. The loopback connection is established out of band. For instance, the application layer protocol may allow the receiver to specify a [BTP/2.0](../0023-bilateral-transfer-protocol/0023-bilateral-transfer-protocol.md) URL for the sender to connect to.
+1. The sender's LT module is initialized with this open connection.
+1. The sender's LT module discovers the loopback address using [IL-DCP](https://github.com/interledgerjs/ilp-protocol-ildcp).
+1. The sender's application layer protocol tells the LT module to send a payment for a given source amount, and expiry time.
+1. The sender's LT module chooses a fulfillment at random.
+1. The sender's LT module creates the condition of their payment by computing the SHA-256 hash of the fulfillment.
+1. The sender's LT module sets the destination address to the loopback address for this lookback link.
+1. Using this condition, address, and amount, the sender's LT module prepares an ILPv4 payment to a connector of her choice.
+1. The payment will find its way from this connector, via any number of additional connectors, to the receiver.
+1. The receiver checks that the destination address matches the loopback address he announced to the sender over IL-DCP earlier, and forwards the ILPv4 `prepare` message to the sender, over the loopback link.
+1. The sender's application layer protocol gets to decide whether the destination amount that looped back is sufficient and whether this particular loopback payment should be accepted or rejected.
+1. The sender's LT module accepts or rejects the ILPv4 payment that looped back over the loopback link.
+1. If for whatever reason the payment does not arrive in time, the LT module responds to the application layer protocol with an error message instead
+
+## How to use
+
+When designing an application layer protocol, it's advisable to support both LT and PSK. Even though LT is far simpler than PSK, the extra loopback hop means that LT payments inherently have a higher latency than PSK payments, where the receiver can immediately fulfill incoming prepared payments. In situations where the sender can easily formulate instructions for the receiver about when to accept a payment and when not, and the sender trusts the receiver to execute these instructions in good faith, the speed up achieved with PSK means money needs to be tied up for less time, and this may translate to lower transaction fees, depending on connector pricing policies. So the application layer protocol needs to provide a choice between LT and PSK.
```

Comment:
> When designing an application layer protocol, it's advisable to support both LT and PSK

I disagree with this statement. Even if there are two options I think it would make much more sense to pick one for your specific use case. On the internet, HTTP doesn't let you use either TCP or UDP.

Reply:
i would see it more like browsers support both http (the simpler, loopback-like protocol) and https (the more complex, psk-like one).

Reply:
If implementations are expected to implement both and one of them is lower-latency, as you describe, why would you use the other?

Reply:
Right, sorry, reading this back, what I meant to say is that a *setup protocol* (like http-ilp, payment-pointers, web-monetization, web-payments, ...) should provide access to both. An application layer protocol in the sense of "the paid audio/video streaming protocol" which is used as a building block for an application, should probably pick one of the two, and then it makes sense to have separately implemented building blocks like that for developers to use.